timemachine/cpp/src/tibd_exchange_move.cu

```diff
@@ -0,0 +1,316 @@
+// if using the same number as in the rest of the kernels of DEFAULT_THREADS_PER_BLOCK
+#define WEIGHT_THREADS_PER_BLOCK 512
+// Currently only support one sample at a time
+#define NUM_SAMPLES 1
+
+namespace timemachine {
+
+template <typename RealType>
+TIBDExchangeMove<RealType>::TIBDExchangeMove(
+    const int N,
+    const std::vector<int> ligand_idxs,
+    const std::vector<std::vector<int>> &target_mols,
+    const std::vector<double> &params,
+    const double temperature,
+    const double nb_beta,
+    const double cutoff,
+    const double radius,
+    const int seed,
+    const int proposals_per_move)
+    : BDExchangeMove<RealType>(N, target_mols, params, temperature, nb_beta, cutoff, seed, proposals_per_move),
+      radius_(static_cast<RealType>(radius)), inner_volume_(static_cast<RealType>((4.0 / 3.0) * M_PI * pow(radius, 3))),
+      d_rand_states_(DEFAULT_THREADS_PER_BLOCK), d_inner_mols_count_(1), d_inner_mols_(this->num_target_mols_),
+      d_outer_mols_count_(1), d_outer_mols_(this->num_target_mols_), d_sorted_indices_(this->num_target_mols_),
+      d_sort_storage_(0), d_center_(3), d_acceptance_(round_up_even(2)), d_targeting_inner_vol_(1),
+      d_ligand_idxs_(ligand_idxs), d_src_weights_(this->num_target_mols_), d_dest_weights_(this->num_target_mols_),
+      d_box_volume_(1), p_inner_count_(1), p_targeting_inner_vol_(1) {
+
+    if (radius <= 0.0) {
+        throw std::runtime_error("radius must be greater than 0.0");
+    }
+
+    // Create event with timings disabled as timings slow down events
+    gpuErrchk(cudaEventCreateWithFlags(&host_copy_event_, cudaEventDisableTiming));
+
+    k_initialize_curand_states<<<1, DEFAULT_THREADS_PER_BLOCK, 0>>>(
+        DEFAULT_THREADS_PER_BLOCK, seed, d_rand_states_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // estimate size needed to do radix sorting
+    // reuse d_sort_keys_in_ rather than constructing a dummy output idxs buffer
+    gpuErrchk(cub::DeviceRadixSort::SortKeys(
+        nullptr, sort_storage_bytes_, d_inner_mols_.data, d_sorted_indices_.data, this->num_target_mols_));
+
+    d_sort_storage_.realloc(sort_storage_bytes_);
+}
+
+template <typename RealType> TIBDExchangeMove<RealType>::~TIBDExchangeMove() {
+    gpuErrchk(cudaEventDestroy(host_copy_event_));
+}
+
+template <typename RealType>
+void TIBDExchangeMove<RealType>::move_device(
+    const int N,
+    double *d_coords, // [N, 3]
+    double *d_box,    // [3, 3]
+    cudaStream_t stream) {
+
+    if (N != this->N_) {
+        throw std::runtime_error("N != N_");
+    }
+
+    // Set the stream for the generator
+    curandErrchk(curandSetStream(this->cr_rng_, stream));
+
+    this->compute_initial_weights(N, d_coords, d_box, stream);
+
+    const int tpb = DEFAULT_THREADS_PER_BLOCK;
+    const int mol_blocks = ceil_divide(this->num_target_mols_, tpb);
+
+    dim3 atom_by_atom_grid(ceil_divide(N, tpb), this->mol_size_, 1);
+
+    k_compute_centroid_of_atoms<RealType>
```

Comment:
i'm guessing we're guaranteeing to never move the ligand during the sequence of moves?

Reply:
Correct